File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == NULL || head->next == NULL) return;

        //Find middle
        ListNode* slow = head;
        ListNode* fast = head;

        while (fast->next != NULL && fast->next->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
        }

        //Reverse second half
        ListNode* curr = slow->next;
        slow->next = NULL;

        ListNode* prev = NULL;
        while (curr) {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }

        // Merge
        ListNode* first = head;
        slow->next = NULL;
        ListNode* second = prev;

        while (second != NULL) {
            ListNode* next1 = first->next;
            ListNode* next2 = second->next;

            first->next = second;   // attach from second list
            second->next = next1;   // reconnect back to first

            first = next1;
            second = next2;
        }
    }
};

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        
        # create a dummy node to handle edge cases (like reversing from head)
        dummy = ListNode(0)
        dummy.next = head

        # pointer to node just before the 'left' position
        leftPre = dummy
        curr = head

        # move leftPre and curr to (left-1) and left positions
        cnt = 0
        while cnt < left - 1:
            cnt += 1
            leftPre = leftPre.next
            curr = curr.next
        
        # mark the start of sublist to be reversed
        sublistHead = curr

        # reverse nodes between left and right
        prev = None
        for _ in range(right - left + 1):
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        
        # reconnect the reversed sublist with the main list
        leftPre.next = prev
        sublistHead.next = curr

        return dummy.next